Implement more array comparisons

stinos opened the pull request:
Two commits:
- fix problems mentioned in #7227, essentially raising an exception whenever someone tries to compare arrays of different types which would previously give incorrect results. So this covers the case where someone accidentally does this (in which case they now know they are doing something wrong) and the case where someone intentionally does this but is unaware that the results are wrong (and the fix is to manually implement the comparison for now - supporting this in the core for all types would be quite the code size hit)
- adding `> < >= <=` which was why I started looking into this: some external piece of code I got uses this to sort bytearrays

stinos committed:
py/objarray: Implement more/less comparisons for array.

stinos commented:
Hmm, I was wondering already why dpgeorge's face turned up for the first commit :) Accidentally did an amend.

dpgeorge commented:
> Hmm, I was wondering already why dpgeorge's face turned up for the first commit :) Accidentally did an amend.

Interesting!  Will you fix that?  I don't mind either way.

The change in code size for this PR is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  +144 +0.028% 
unix nanbox:  +140 +0.031% 
      stm32:   +72 +0.019% PYBV10
     cc3200:   +64 +0.035% 
    esp8266:  +112 +0.016% GENERIC
      esp32:  +100 +0.007% GENERIC
        nrf:   +76 +0.051% pca10040
        rp2:   +80 +0.017% PICO
       samd:   +76 +0.074% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
IMO the change is needed because otherwise it's too easy to let incorrect code silently pass.

dpgeorge commented on py/objarray.c:
```diff
@@ -258,6 +258,19 @@ STATIC mp_obj_t array_unary_op(mp_unary_op_t op, mp_obj_t o_in) {
+    }
+    if (typecode == 'f' || typecode == 'd') {
+        return typecode;
+    }
```
I think you can just remove this if statement completely, it won't be caught by the if below and should be equivalent.  My testing shows that saves a handful of bytes.

stinos replied:
Without this the typecode always becomes uppercase, so the `lhs_code != 'f' && lhs_code != 'd'` below are never true and we'd be comparing floating point arrays. I can go back to my first attempt though, which removes this bit and below compares `lhs_code != 'F' && lhs_code != 'D'`. Works, but is just somewhat confusing because `F` and `D` aren't valid typecodes anywhere. But I'd add a comment to explain it.

stinos replied:
Changed this, added a comment, and removed the duplicate comparison against 'Z' to save extra bytes. For unix this is now +96 instead of +144, of which +64 is for the first commit.

dpgeorge replied:
Nice little improvement!

Could the function always return lowercase instead of uppercase?  Then it's more obvious with the f/d.

stinos replied:
Yes but that needs removing the early return in case of BYTEARRAY_TYPECODE and using `is_unsigned` instead to be able to keep the same amount/type of instructions. Not really a problem, will do so right away.

dpgeorge commented:
Good set of new tests!

stinos committed:
py/objarray: Prohibit comparison of mismatching types.

Array equality is defined as each element being equal but to keep
code size down MicroPython implements a binary comparison.  This
can only be used correctly for elements with the same binary layout
though so turn it into an NotImplementedError when comparing types
for which the binary comparison yielded incorrect results: types
with different sizes, and floating point numbers because nan != nan.

dpgeorge commented:
Size change now:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +96 +0.019% 
unix nanbox:   +96 +0.021% 
      stm32:   +56 +0.014% PYBV10
     cc3200:   +48 +0.026% 
    esp8266:   +96 +0.014% GENERIC
      esp32:   +68 +0.005% GENERIC
        nrf:   +60 +0.041% pca10040
        rp2:   +56 +0.012% PICO
       samd:   +60 +0.058% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

dpgeorge commented:
Looks really good now!

dpgeorge commented:
Merged in 57365d855734142deb030ebcd00c10efcedf554b and 09be0c083cb493c70b38079aa6fd785aa9b0b90a
